File: chisel_js/src/chisel/checker/services/svelte/complexity.ts
import type { ProjectInfo } from "chisel/checker/models/project_info";
import type { Violation } from "chisel/checker/models/violation";
import { Severity } from "chisel/checker/models/severity";
import { createViolation } from "chisel/checker/models/violation";

export class ComplexityService {
  readonly ruleIdPrefix = "complexity";

  check(project: ProjectInfo): Violation[] {
    const violations: Violation[] = [];
    for (const file of project.files) {
      if (!file.source) continue;
      violations.push(...this._checkLoc(file));
    }
    return violations;
  }

  private _checkLoc(file: { path: string; source: string; layer: string }) {
    const violations: Violation[] = [];
    const lines = file.source.split("\n");
    const loc = lines.filter(l => l.trim() && !l.trim().startsWith("//")).length;

    if (file.path.endsWith("+page.svelte")) {
      if (loc > 100) {
        violations.push(createViolation({
          file: file.path, line: 1, severity: Severity.ERROR,
          ruleId: "complexity:page-loc-limit",
          message: `+page.svelte exceeds 100 lines of code (found ${loc}). Extract sections into components/domain/.`,
        }));
      } else if (loc > 80) {
        violations.push(createViolation({
          file: file.path, line: 1, severity: Severity.WARNING,
          ruleId: "complexity:page-loc-warning",
          message: `+page.svelte exceeds 80 lines of code (found ${loc}). Consider extracting sections into components/domain/.`,
        }));
      }
    }

    if (file.path.includes("controllers/") && file.path.endsWith(".ts")) {
      const methodMatches = file.source.matchAll(/(?:async\s+)?\w+\s*\([^)]*\)\s*(?::\s*\w+)?\s*\{/g);
      for (const m of methodMatches) {
        const startLine = file.source.substring(0, m.index!).split("\n").length;
        let braceCount = 0;
        let endLine = startLine;
        let started = false;
        const remaining = file.source.substring(m.index!);
        for (let j = 0; j < remaining.length; j++) {
          if (remaining[j] === '{') { braceCount++; started = true; }
          if (remaining[j] === '}') {
            braceCount--;
            if (started && braceCount === 0) {
              endLine = file.source.substring(0, m.index! + j).split("\n").length;
              break;
            }
          }
        }
        const methodLoc = endLine - startLine + 1;
        if (methodLoc > 40) {
          violations.push(createViolation({
            file: file.path, line: startLine, severity: Severity.ERROR,
            ruleId: "complexity:controller-loc-limit",
            message: `Controller method exceeds 40 lines of code (found ${methodLoc}). Extract business logic into a service.`,
          }));
        }
      }
    }

    if ((file.path.endsWith("+page.server.ts") || file.path.endsWith("+layout.server.ts"))) {
      const funcMatches = file.source.matchAll(/(?:export\s+(?:async\s+)?function\s+(load|actions))\s*\([^)]*\)\s*(?::\s*\w+)?\s*\{/g);
      for (const m of funcMatches) {
        const startLine = file.source.substring(0, m.index!).split("\n").length;
        let braceCount = 0;
        let endLine = startLine;
        let started = false;
        const remaining = file.source.substring(m.index!);
        for (let j = 0; j < remaining.length; j++) {
          if (remaining[j] === '{') { braceCount++; started = true; }
          if (remaining[j] === '}') {
            braceCount--;
            if (started && braceCount === 0) {
              endLine = file.source.substring(0, m.index! + j).split("\n").length;
              break;
            }
          }
        }
        const funcLoc = endLine - startLine + 1;
        if (funcLoc > 20) {
          violations.push(createViolation({
            file: file.path, line: startLine, severity: Severity.ERROR,
            ruleId: "complexity:loader-loc-limit",
            message: `Loader/action exceeds 20 lines of code (found ${funcLoc}). Move logic into a controller or service.`,
          }));
        }
      }
    }

    return violations;
  }

  describeRules() {
    return [
      { id: "complexity:page-loc-limit", category: "complexity",
        description: "+page.svelte exceeds 100 LoC", fixGuidance: "Extract logical sections into components/domain/." },
      { id: "complexity:page-loc-warning", category: "complexity",
        description: "+page.svelte exceeds 80 LoC", fixGuidance: "Consider extracting sections into components/domain/." },
      { id: "complexity:controller-loc-limit", category: "complexity",
        description: "Controller method exceeds 40 LoC", fixGuidance: "Extract business logic into a service. Controllers orchestrate — services do the work." },
      { id: "complexity:loader-loc-limit", category: "complexity",
        description: "Loader or form action exceeds 20 LoC", fixGuidance: "Move logic into a controller or service. Loaders parse input, call factory, return output." },
    ];
  }
}

File: src/chisel/checker/services/complexity.py
from __future__ import annotations

import ast
from dataclasses import dataclass

from chisel.checker.models.file_info import FileInfo
from chisel.checker.models.layer import Layer
from chisel.checker.models.project_info import ProjectInfo
from chisel.checker.models.severity import Severity
from chisel.checker.models.violation import Violation

_CONTROLLER_MAX_LOC = 30
_CONTROLLER_MAX_COMPLEXITY = 3
_ROUTE_MAX_LOC = 20
_APP_MAX_LOC = 50


@dataclass(slots=True)
class ComplexityService:
    rule_id_prefix: str = "complexity"

    def check(self, project: ProjectInfo) -> list[Violation]:
        violations: list[Violation] = []
        for file in project.files:
            if file.source == "" or file.ast_tree is None:
                continue
            violations.extend(self._check_file(file))
        return violations

    def _check_file(self, file: FileInfo) -> list[Violation]:
        violations: list[Violation] = []
        violations.extend(self._check_loc(file))
        violations.extend(self._check_method_complexity(file))
        violations.extend(self._check_factory_complexity(file))
        return violations

    def _check_loc(self, file: FileInfo) -> list[Violation]:
        lines = file.source.split("\n")
        loc = len([l for l in lines if l.strip() and not l.strip().startswith("#")])

        if file.layer == Layer.APP_FILE and loc > _APP_MAX_LOC:
            return self._v(
                file, 1, "app-loc-limit",
                f"app.py must be ≤ {_APP_MAX_LOC} lines of code (found {loc})",
            )

        if file.layer == Layer.ROUTES:
            violations: list[Violation] = []
            tree = file.ast_tree
            if tree is None:
                return violations
            for node in tree.body:
                if isinstance(node, ast.FunctionDef):
                    end = node.end_lineno or node.lineno
                    func_loc = end - node.lineno + 1
                    if func_loc > _ROUTE_MAX_LOC:
                        violations.extend(
                            self._v(
                                file, node.lineno, "route-loc-limit",
                                f"Route endpoint '{node.name}' must be "
                                f"≤ {_ROUTE_MAX_LOC} lines of code "
                                f"(found {func_loc})",
                            )
                        )
            return violations

        if file.layer == Layer.CONTROLLERS:
            violations: list[Violation] = []
            tree = file.ast_tree
            if tree is None:
                return violations
            for node in tree.body:
                if isinstance(node, ast.ClassDef):
                    for item in node.body:
                        if isinstance(item, ast.FunctionDef):
                            end = item.end_lineno or item.lineno
                            func_loc = end - item.lineno + 1
                            if func_loc > _CONTROLLER_MAX_LOC:
                                violations.extend(
                                    self._v(
                                        file, item.lineno, "controller-loc-limit",
                                        f"Controller method '{node.name}."
                                        f"{item.name}' must be "
                                        f"≤ {_CONTROLLER_MAX_LOC} lines of code "
                                        f"(found {func_loc})",
                                    )
                                )
            return violations

        return []

    def _check_method_complexity(self, file: FileInfo) -> list[Violation]:
        if file.layer != Layer.CONTROLLERS:
            return []

        violations: list[Violation] = []
        tree = file.ast_tree
        if tree is None:
            return violations

        for node in tree.body:
            if isinstance(node, ast.ClassDef):
                for item in node.body:
                    if isinstance(item, ast.FunctionDef):
                        cc = self._cyclomatic_complexity(item)
                        if cc > _CONTROLLER_MAX_COMPLEXITY:
                            violations.extend(
                                self._v(
                                    file, item.lineno,
                                    "controller-complexity-limit",
                                    f"Controller method '{node.name}."
                                    f"{item.name}' has cyclomatic complexity "
                                    f"{cc} (max {_CONTROLLER_MAX_COMPLEXITY})",
                                )
                            )
        return violations

    def _check_factory_complexity(self, file: FileInfo) -> list[Violation]:
        if file.layer != Layer.FACTORY:
            return []

        tree = file.ast_tree
        if tree is None:
            return []

        cc = self._file_complexity(tree)
        if cc > 1:
            return self._v(
                file, 1, "factory-complexity-limit",
                f"Factory cyclomatic complexity must be 1 (found {cc})",
            )
        return []

    def _cyclomatic_complexity(self, node: ast.FunctionDef) -> int:
        complexity = 1
        for child in ast.walk(node):
            if isinstance(child, (ast.If, ast.For, ast.While, ast.ExceptHandler)):
                complexity += 1
            elif isinstance(child, ast.BoolOp):
                complexity += len(child.values) - 1
        return complexity

    def _file_complexity(self, tree: ast.Module) -> int:
        complexity = 1
        for node in ast.walk(tree):
            if isinstance(node, (ast.If, ast.For, ast.While, ast.ExceptHandler)):
                complexity += 1
            elif isinstance(node, ast.BoolOp):
                complexity += len(node.values) - 1
        return complexity

    def _v(
        self, file: FileInfo, line: int, rule_suffix: str, message: str
    ) -> list[Violation]:
        return [
            Violation(
                file=str(file.path),
                line=line,
                severity=Severity.ERROR,
                rule_id=f"{self.rule_id_prefix}:{rule_suffix}",
                message=message,
            )
        ]
